hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionLog.java

```diff
@@ -38,53 +37,40 @@
  * the TransactionLog will contains 4 {@link TransactionLogRecord} instances.
  *
  * planned optimization:
- * Most transaction will be small, but an linkedlist + hashmap is created. Instead use an array and do a
+ * Most transaction will be small, but an linkedhashmap is created. Instead use an array and do a
  * linear search in that array. When there are too many items added, then enable the hashmap.
  */
 public class TransactionLog {
 
-    private final List<TransactionLogRecord> recordList = new LinkedList<TransactionLogRecord>();
-    private final Map<Object, TransactionLogRecord> recordMap = new HashMap<Object, TransactionLogRecord>();
+    private final Map<Object, TransactionLogRecord> recordMap = new LinkedHashMap<>();
 
     public TransactionLog() {
     }
 
     public TransactionLog(List<TransactionLogRecord> transactionLog) {
-        //
-        recordList.addAll(transactionLog);
+        for (TransactionLogRecord record : transactionLog) {
+            add(record);
+        }
     }
 
     public void add(TransactionLogRecord record) {
         Object key = record.getKey();
-
-        // there should be just one tx log for the same key. so if there is older we are removing it
-        if (key != null) {
-            TransactionLogRecord removed = recordMap.remove(key);
-            if (removed != null) {
-                recordList.remove(removed);
-            }
-        }
-
-        recordList.add(record);
-
-        if (key != null) {
-            recordMap.put(key, record);
+        if (key == null) {
+            key = new Object();
         }
```

Comment:
An alternative, maybe better implementation, would be to change semantics of `TransactionLogRecord.getKey()` method and dissallow returning nulls from this method. Then a new field `key` should be added to `ClusterStateTransactionLogRecord`, it's value will be `new Object()` and it will be set in the constructor of this class.

Reply:
I was thinking the same and we can, in fact, change the behaviour in 4.0 but you can add this in a separate PR.

Reply:
OK, will do in a separate PR later.